python/lsst/cp/pipe/defects.py

```diff
@@ -0,0 +1,645 @@
+        validateIsrConfig(self.isrForFlats, self.config.isrMandatoryStepsFlats,
+                          self.config.isrForbiddenStepsFlats, self.config.isrDesirableSteps)
+        validateIsrConfig(self.isrForDarks, self.config.isrMandatoryStepsDarks,
+                          self.config.isrForbiddenStepsDarks, self.config.isrDesirableSteps)
+        self.config.validate()
+        self.config.freeze()
+
+    @classmethod
+    def _makeArgumentParser(cls):
+        """Augment argument parser for the FindDefectsTask."""
+        parser = pipeBase.ArgumentParser(name=cls._DefaultName)
+        parser.add_argument("--visitList", dest="visitList", nargs="*",
+                            help="List of visits to use. Same for each detector." +
+                            " Uses the normal 0..10:3^234 syntax")
+        parser.add_id_argument("--id", datasetType="newDefects",
+                               ContainerClass=NonexistentDatasetTaskDataIdContainer,
+                               help="The ccds to use, e.g. --id ccd=0..100")
+        return parser
+
+    @pipeBase.timeMethod
+    def runDataRef(self, dataRef, visitList):
+        """Run the defect finding task
+
+        TODO: Write docstring
+
+        Parameters
+        ----------
+        dataRef : list of lsst.daf.persistence.ButlerDataRef
+            dataRef for the detector for the visits to be fit.
+        visitList : `iterable` of `int`
+            List of visit numbers to be processed
+        """
+
+        detNum = dataRef.dataId[self.config.ccdKey]
+        self.log.info(f"Calculating defects using {visitList} visits for detector {detNum}")
+
+        defectListFromDarks = []
+        defectListFromFlats = []
+
+        # TODO: Pull polarity handling out here so that the defectTypes config can work
+
+        if self.config.mode == 'MASTER':
+            if len(visitList) > 1:
+                raise RuntimeError(f"Must only specify one visit when using mode MASTER, got {visitList}")
+            dataRef.dataId['visit'] = visitList[0]
+
+            exp = dataRef.get('dark')
+            defects = self.findHotAndColdPixels(exp, 'dark')
+            msg = f"Found {len(defects)} defects containing {self._nPixFromDefects(defects)} pixels"
+            msg += " in master dark"
+            self.log.info(msg)
+            defectListFromDarks.append(defects)
+            if self.config.makePlots:
+                self._plot(dataRef, exp, visitList[0], self.config.nSigmaDark, defects, 'dark')
+
+            exp = dataRef.get('flat')
+            defects = self.findHotAndColdPixels(exp, 'flat')
+            msg = f"Found {len(defects)} defects containing {self._nPixFromDefects(defects)} pixels"
+            msg += " in master flat"
+            self.log.info(msg)
+            defectListFromFlats.append(defects)
+            if self.config.makePlots:
+                self._plot(dataRef, exp, visitList[0], self.config.nSigmaBright, defects, 'flat')
```

Comment:
Again, I think that the workhorse code being factored into `findHotAndColdPixels()` is already at the right level. How strongly do you feel about this?

Reply:
Naïvely, it seems like it would be pretty easy to hit a middle-ground that didn't generate more functions but which reduced the repetitition a bit. Something like

```python
defectLists = {'dark': [], 'flat': []}
for datasetType in defectLists.keys():
    exp = dataRef.get(datasetType)
    defects = self.findHotAndColdPixels(exp, datasetType)

    # NB, probably shouldn't actually be using string interpolation here, but sending direct to the logger
    msg = f"Found {len(defects)} defects containing {self._nPixFromDefects(defects)} pixels in master {datasetType}"  
    self.log.info(msg)

    defectLists[datasetType].append(defects)
    if self.config.makePlots:
        self._plot(dataRef, exp, visitList[0], self.config.nSigmaDark, defects, datasetType)
```
feels like the right level of refactoring to me (obviously, untested, E&OE, etc).

Reply:
Nice suggestion, thanks. Have tweaked and done that. Also note that these changes are happening on DM-19709 (as discussed) so things like the `f`-strings are fixed over there already.